core-rust/state-manager/src/store/proofs_gc.rs

```diff
@@ -0,0 +1,223 @@
+ * permissions under this License.
+ */
+
+use radix_engine::types::{Categorize, Decode, Encode};
+
+use node_common::config::limits::{
+    MAX_TXNS_FOR_RESPONSES_SPANNING_MORE_THAN_ONE_PROOF, MAX_TXN_BYTES_FOR_A_SINGLE_RESPONSE,
+};
+use radix_engine_common::types::Epoch;
+use std::sync::Arc;
+use std::time::Duration;
+use tracing::{error, info};
+
+use crate::store::traits::gc::{
+    LedgerProofsGcProgress, LedgerProofsGcProgressV1, LedgerProofsGcStore,
+};
+use crate::store::traits::proofs::QueryableProofStore;
+
+use crate::store::StateManagerDatabase;
+
+use node_common::locks::StateLock;
+
+/// A configuration for [`LedgerProofsGc`].
+#[derive(Debug, Categorize, Encode, Decode, Clone, Default)]
+pub struct LedgerProofsGcConfig {
+    /// How often to run the GC, in seconds.
+    /// Since this GC operates with an epoch precision, we do not need to run more often than epoch
+    /// changes.
+    // TODO(after having some event-driven Rust infra): The entire `LedgerProofsGc` could be
+    // migrated away from `Scheduler` into `EventListener<EpochChangeCommittedEvent>` (as noted
+    // above - it wants to run async exactly once after each epoch).
+    pub interval_sec: u32,
+    /// How many most recent *completed* epochs should be left not GC-ed.
+    /// Please note that the current epoch is never GC-ed.
+    pub most_recent_full_resolution_epoch_count: usize,
+}
+
+/// A garbage collector of sufficiently-old, non-critical ledger proofs.
+/// A ledger proof is "non-critical" when it is not listed by our "get transactions with their
+/// proof" logic (used e.g. for ledger-sync responses).
+/// The implementation is suited for being driven by an external scheduler.
+pub struct LedgerProofsGc {
+    database: Arc<StateLock<StateManagerDatabase>>,
+    interval: Duration,
+    most_recent_full_resolution_epoch_count: u64,
+}
+
+impl LedgerProofsGc {
+    /// Creates a new GC.
+    pub fn new(
+        database: Arc<StateLock<StateManagerDatabase>>,
+        config: LedgerProofsGcConfig,
+    ) -> Self {
+        Self {
+            database,
+            interval: Duration::from_secs(u64::from(config.interval_sec)),
+            most_recent_full_resolution_epoch_count: u64::try_from(
+                config.most_recent_full_resolution_epoch_count,
+            )
+            .unwrap(),
+        }
+    }
+
+    /// An interval between [`run()`]s, to be used by this instance's scheduler.
+    pub fn interval(&self) -> Duration {
+        self.interval
+    }
+
+    /// Performs a single GC run, which is supposed to permanently delete *all* non-critical ledger
+    /// proofs of configured old-enough epochs.
+    pub fn run(&self) {
+        // TODO(locks/snapshots): The GC's operation does not interact with the "current state", and
+        // intuitively could use the "historical, non-locked" DB access. However, we have a (very
+        // related) use-case, which lists arbitrary past transactions and their proof - and it could
```

Comment:
Can we be more explicit? You're referring to ledger sync here I believe?

---

Also I'm wondering about if we could just have a `ledger_sync` lock, which is taken out as readable to serve ledger sync responses, and is taken out as writable by this process? (I think ledger extension doesn't need to pick up this lock as it's atomic - although we'd want to check that the `get_txns_and_proof` doesn't assume any such invariants).

This also decreases contention whereby ledger sync requests could slow down commit/consensus 👍

Reply:
> if we could just have a `ledger_sync` lock

Yes, that's a good direction, but I don't want to go there in this PR.

`<off-topic>`:
I think that the current shape of `StateLock` (our DB lock) clearly shows that we don't want to lock a single "DB the resource", but rather a set of abstractly defined "accesses to DB". E.g.:

- Lock 1: 
  - Access: "Consistent current top-of-ledger"
  - Typical reader: "get current state version and values of a couple substates" (i.e. state-reading Core API endpoints)
  - Typical writer: Ledger Commit

- Lock 2:
  - Access: "A sequence of transactions and their end-proof"
  - Typical reader: Ledger Sync responder
  - Typical writer: Ledger Proofs GC

_(possibly more)_

And the `&database` itself is simply always accessible without any lock. This means that readers/writers would have to be a lot more careful... (remember about locking R or W of the lock appropriate to your Access + be careful to never interact with the parts of DB guarded by locks of other Accesses).

TBH I'd refactor it right away, but I acknowledge the same area is touched by @malexandru-rdx's "db snapshots" work, so let's first see that.
`</off-topic>`

So, in this PR, let me (fix and) leave the full R/W-locking of entire DB for GC purposes.
(GC process organically releases & re-acquires the lock, so even during the backlog it should not starve the BFT process)